Fix  #2180

harsha544 opened the pull request:
Fix #2180 

Updating relevant package.json  and karma.conf.js files to pick up karma-firefox-launcher
and run relevant UT(s) in Firefox's headless mode. 

The need for firefox arouse because of it's availability across all arch.

Signed-off-by: Krishna Harsha Voora <krishvoor@in.ibm.com>

jt-nti commented:
As there any background to the PR? (It's not linked to an issue.) Also, it's not working for me...

```
20 09 2017 13:38:43.799:INFO [launcher]: Starting browser Firefox
20 09 2017 13:38:43.802:ERROR [launcher]: No binary for Firefox browser on your platform.
  Please, set "FIREFOX_BIN" env variable.
```

Are there any new setup steps for a development environment that need documenting?

harsha544 commented:
I have now linked the issue to the commit message which is #2180 

Can you try set FIREFOX_BIN variable via `export FIREFOX_BIN=/usr/bin/firefox` assuming you have firefox is installed and accessible via /usr/bin/firefox ?

jt-nti commented:
I don't have Firefox installed. Could you update more of the development environment documentation/setup scripts to match the changes please. For example...

- https://github.com/hyperledger/composer/blob/master/contrib-notes/getting-started.md
- https://github.com/hyperledger/composer/blob/master/contrib-notes/prerequisites.md
- https://github.com/hyperledger/composer/blob/master/packages/composer-website/jekylldocs/prereqs-ubuntu.sh

It would also be good to include custom preferences to configure the tests to run headlessly, until karma-firefox-launcher comes with that support as standard:

karma-runner/karma-firefox-launcher#76

Thanks

jt-nti commented:
Just spotted that this change will also mean that testing locally will be on Firefox, whereas Travis builds will still be using Chrome which doesn't seem right to me

harsha544 commented:
Will update the commit to include the changes to
https://github.com/hyperledger/composer/blob/master/packages/composer-website/jekylldocs/prereqs-ubuntu.sh 
https://github.com/hyperledger/composer/blob/master/contrib-notes/getting-started.md#tool-chain-reference-details

And I see Travis CI Checks for this PR making use of firefox instead of Chrome.

harsha544 commented:
The only reason we are switching to Firefox is , Chrome/Chromium is not currently available for ppc64le arch, so if we include ppc64le into CI environment we may not have successful builds at all which would be a blocker for publishing the relevant docker-images for all architectures.

I'm a newbie at JS so will take sometime before i push changes corresponding to headless firefox.

harsha544 commented:
To summarize I have made relevant changes to JS which now should trigger Firefox in headless mode and run relevant UT(s) in the same.

jt-nti commented on packages/composer-playground/config/karma.conf.js:
```diff
@@ -76,10 +76,13 @@ module.exports = function(config) {
-    ],
-
+    browsers: ['FirefoxHeadless'],
+    customLaunchers: {
```
This is getting replaced below, resulting in the error: `Cannot load browser "FirefoxHeadless": it is not registered!`

harsha544 replied:
Firefox headless option works from Version 55+ for linux per this https://developer.mozilla.org/en-US/Firefox/Headless_mode . And I see Travis CI Nodes is already equipped with Firefox 55+
Request for a re-review with new changes in.

jt-nti replied:
Unfortunately version 56+ is required on Windows/Mac, which I think is what the majority of developers are using. Hopefully not long to wait though... https://wiki.mozilla.org/RapidRelease/Calendar

harsha544 commented:
I obtained the following from Travis CI build logs. 

  76:11  error  Expected indentation of 12 spaces but found 10  indent
  77:13  error  Expected indentation of 14 spaces but found 12  indent
  78:13  error  Expected indentation of 14 spaces but found 12  indent

Before updating my patch I fixed the above using **eslint karma.conf.js --fix** .
Apparently that didn't fix.  Any pointers on how to address the above ?

jt-nti commented on packages/composer-playground/config/karma.conf.js:
```diff
@@ -105,7 +101,7 @@ module.exports = function(config) {
 
   if (process.env.TRAVIS){
```
Doesn't look like this is necessary. configuration.browsers should already be set up for firefox above

harsha544 commented:
I need pointers on how to fix this issue :- 
I obtained the following from Travis CI build logs.

76:11 error Expected indentation of 12 spaces but found 10 indent
77:13 error Expected indentation of 14 spaces but found 12 indent
78:13 error Expected indentation of 14 spaces but found 12 indent

Before updating my patch I fixed the above using eslint karma.conf.js --fix .
Apparently that didn't fix.

harsha544 commented:
@jt-nti I have messed my local git repo for composer, hence opened up https://github.com/hyperledger/composer/pull/2238 where I addressed your comments. Request for review there. Thank you !!
